tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -7,125 +7,434 @@
+                    var value = header.Substring(colonIndex + 1, endIndex - colonIndex);
 
-                spanContext = new SpanContext(traceId, parentId, samplingPriority, serviceName: null, null, w3cTraceId, w3cSpanId);
+                    if (name == "s")
+                    {
+                        // SamplingPriorityToInt32(string)
+                        samplingPriority = SamplingPriorityToInt32(value);
+                    }
+                    else if (name == "o")
+                    {
+                        origin = value;
+                    }
+                    else if (name.StartsWith("t.", StringComparison.Ordinal))
+                    {
+                        propagatedTagsBuilder.Append("_dd.p.").Append(name.Substring(2)).Append('=').Append(value).Append(',');
+                    }
 #endif
 
+                    // skip past the semicolon
+                    startIndex = endIndex + 2;
+                }
+
+                string? propagatedTags;
+
+                if (propagatedTagsBuilder.Length > 0)
+                {
+                    if (propagatedTagsBuilder[propagatedTagsBuilder.Length - 1] == ',')
+                    {
+                        propagatedTagsBuilder.Length--;
+                    }
+
+                    propagatedTags = StringBuilderCache.GetStringAndRelease(propagatedTagsBuilder);
+                }
+                else
+                {
+                    propagatedTags = null;
+                }
+
+                traceState = new W3CTraceState(samplingPriority, origin, propagatedTags);
                 return true;
             }
+            finally
+            {
+                StringBuilderCache.Release(propagatedTagsBuilder);
+            }
+        }
+
+        [return: NotNullIfNotNull("samplingPriority")]
+        private static string? SamplingPriorityToString(int? samplingPriority)
+        {
+            return samplingPriority switch
+                   {
+                       2 => "2",
+                       1 => "1",
+                       0 => "0",
+                       -1 => "-1",
+                       null => null,
+                       not null => samplingPriority.Value.ToString(CultureInfo.InvariantCulture)
+                   };
+        }
 
-            return false;
+        private static int? SamplingPriorityToInt32(string? samplingPriority)
+        {
+            return samplingPriority switch
+                   {
+                       "2" => 2,
+                       "1" => 1,
+                       "0" => 0,
+                       "-1" => 1,
+                       null => null,
+                       not null => int.TryParse(samplingPriority, out var result) ? result : null
+                   };
         }
 
-        private static bool IsValidTraceId([NotNullWhen(true)] string? traceId)
+#if NETCOREAPP
+        private static int? SamplingPriorityToInt32(ReadOnlySpan<char> samplingPriority)
```

Comment:
FYI, I think in C# 10 you can implement this identically to the `string` version:

```csharp
        private static int? SamplingPriorityToInt32(ReadOnlySpan<char> samplingPriority)
          => samplingPriority switch
             {
                "2" => 2,
                "1" => 1,
                "0" => 0,
                "-1" => 1,
                _ => int.TryParse(samplingPriority, out var result) ? result : null
             };
```

Reply:
Interesting! But it's C# 11 which Rider doesn't seem to support yet 😭

Reply:
> But it's C# 11 which Rider doesn't seem to support yet

Use the EAP, that's supported! 😉 And it's been stable enough for me 🙂

Reply:
Rider 2022.3 was released in the last hour or two!

Reply:
Came here to post this link and see you beat me to it 😄 https://www.jetbrains.com/rider/whatsnew/

Reply:
Fixed.